### PR TITLE
Remove Virtus dependency

### PR DESCRIPTION
**Pull request by jturkel**
This PR prototypes removing the [virtus](https://github.com/solnic/virtus) dependency since that gem is slow and no longer maintained. I took a look at [dry-struct](https://dry-rb.org/gems/dry-struct/)  (the virtus replacement) but that gem seemed like overkill and it still resulted in complicated code because the attribute type system doesn't know anything about the Avro type system. 

The big changes in behavior are:
* Coercion is much more strict (e.g. we don't try to coerce a string to an integer) and throws `ArgumentError`s if something can't be coerced. I believe this should be fine since Avromatic models should be constructed off already sanitized data rather than user input. This leads to faster to detection of bugs and improved performance.
* Unions with custom types are now supported (although there may be edge cases that I'm not thinking of)

Here are the benchmark results for a `User` model (a simple model with only a few attributes) and a `Product` model (a complex model with lots of fields and unions) that show nice speeds for constructing model instances and serializing models to Avro:

```
Product Construction:
Original     222.338  (±12.1%) i/s -      6.573k in  30.017785s
Optimized    640.343  (±10.0%) i/s -     19.008k in  30.015480s

User Construction:
Original    30.914k (± 6.6%) i/s -    923.824k in  30.024677s
Optimized   89.635k (± 8.8%) i/s -      2.674M in  30.076136s

Product Messaging Serialization
Original     89.355  (± 9.0%) i/s -      2.660k in  30.014785s
Optimized   181.148  (± 9.9%) i/s -      5.392k in  30.081897s

User Messaging Serialization
Original     5.893k (± 7.1%) i/s -    175.929k in  30.021871s
Optimized    7.998k (±10.7%) i/s -    237.402k in  30.040313s
```
The high level design involves:
* Introducing a hierarchy of types under `Avromatic::Model::Types` for coercing a value to the appropriate type and serializing values back to an appropriate Avro datum
* Storing a hash of `AttributeDefinition`s on each model that includes details on the attribute like the name, default and type

Things I'd love feedback on at this stage:
* Do you anticipate any problems with getting stricter about our coercion rules? 
* Does the modeling of types and attributes make sense?
* How should we handle unknown keys passed to a model constructor? My gut is we should raise an exception since it's a likely indicator of a bug in client code.
* How to more efficiently implement `Avromatic::Model::Types::CustomType` without complicating the custom type configuration experience.
* Any other high level issues you see with this approach.

### CODE REVIEW

File: lib/avromatic/patches/schema_validator_patch.rb [hunk 1]
```diff
@@ -3,8 +3,8 @@ module Patches
     module SchemaValidatorPatch
       # This method replaces validate_recursive in AvroPatches::LogicalTypes::SchemaValidatorPatch
       # to enable validating datums that contain an encoding provider.
-      def validate_recursive(expected_schema, logical_datum, path, result, encoded = false)
-        datum = resolve_datum(expected_schema, logical_datum, encoded)
+      def validate_recursive(expected_schema, logical_datum, path, result, options = { encoded: false })
```
Comment (jturkel) [inline lib/avromatic/patches/schema_validator_patch.rb]: This is temporary until https://github.com/salsify/avromatic/pull/85 lands.

File: lib/avromatic/model/types/array_type.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        def coerced?(value)
+          value.nil? || (value.is_a?(::Array) && value.all? { |element| value_type.coerced?(element) })
+        end
+
+        def serialize(value, strict:)
+          if value.nil?
+            value
+          else
+            value.map { |element| value_type.serialize(element, strict: strict) }
```
Comment (jkapell) [inline lib/avromatic/model/types/array_type.rb]: Use safe-nav operator and skip nil check?

File: lib/avromatic/model/types/boolean_type.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    module Types
+      class BooleanType
+        VALUE_CLASSES = [::TrueClass, ::FalseClass].freeze
+
+        def value_classes
+          VALUE_CLASSES
+        end
+
+        def coerce(input)
+          if input.nil? || input.is_a?(::TrueClass) || input.is_a?(::FalseClass)
```
Comment (jkapell) [inline lib/avromatic/model/types/boolean_type.rb]: Could just call `coercible?` here.

File: lib/avromatic/model/types/type_factory.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            when 'timestamp-micros'
+              return Avromatic::Model::Types::TimestampMicrosType.new
+            when 'timestamp-millis'
+              return Avromatic::Model::Types::TimestampMillisType.new
+            end
+          end
+
+          case schema.type_sym
+          when :string, :bytes
+            Avromatic::Model::Types::StringType.new
```
[the remark this answers is not in the document]
Comment (jturkel) [reply lib/avromatic/model/types/type_factory.rb]: They could be static but I went with instances to keep all of the Type classes consistent. I can reduce memory usage a bit and simplify the conditional code a bit if I stick the simple types into a static hash.

File: lib/avromatic/model/types/type_factory.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            Avromatic::Model::Types::ArrayType.new(value_type: value_type)
+          when :map
+            value_type = create(schema: schema.values, nested_models: nested_models, use_custom_types: use_custom_types)
+            Avromatic::Model::Types::MapType.new(
+              key_type: Avromatic::Model::Types::StringType.new,
+              value_type: value_type
+            )
+          when :union
+            null_index = schema.schemas.index { |member_schema| member_schema.type_sym == :null }
+            raise 'a null type in a union must be the first member' if null_index && null_index > 0
```
[the remark this answers is not in the document]
Comment (jturkel) [reply lib/avromatic/model/types/type_factory.rb]: Agreed. I'll investigate that separately before releasing 2.0 since this is already a big change.

File: lib/avromatic/model/types/custom_type.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+        def coercible?(input)
+          # TODO: Is there a better way to implement this?
+          input.nil? || !coerce(input).nil?
+        rescue
+          false
+        end
+
+        def coerced?(value)
+          # TODO: Is there a better way to implement this?
```
[the remark this answers is not in the document]
Comment (jturkel) [reply lib/avromatic/model/types/custom_type.rb]: Sounds good. This PR is already pretty big so I'm going to work on getting it landed in its current form and then I'll put up a separate PR for custom type configuration of `coerced?` / `coercible?`.

**Comment by jturkel** [comment]
@tjwp / @jkapell - I pushed some updates and raised a few questions in my responses when you have time to take a look. I'd like to get this landed and then follow up with a few smaller PRs before releasing a 2.0 alpha:

1. More informative coercion errors that include context information like the schema field name
1. Automatic validation when serialization to Avro for better error messages
1. Raise exception for unknown attributes passed to model constructors
1. CustomType configuration for efficient `coercible?` and `coerced?` checks
1. A more detailed 2.0 upgrade guide
1. A Union coercion algorithm that looks at all attributes of input for more accurate matching (this may fall out naturally from raising an exception for constructing a model with unknown attributes)
1. Unions with nulls that are not the first element (although I might punt on this since we don't have an immediate need for it)

**Comment by jturkel** [comment]
I added more test coverage but hit the point of diminishing returns and lowered the minimum test coverage bar to 95% for now.

File: lib/avromatic/model/attributes.rb [hunk 7]
```diff
@@ -20,16 +20,54 @@ def initialize(field)
...
-        # This hack lives on for now because custom type coercion is not pushed
-        # down into unions. This means that custom types can only be optional
-        # fields, not members of real unions.
-        field_type.schemas.reject { |schema| schema.type_sym == :null }.first
+      class AttributeDefinition
+        attr_reader :name, :type, :field, :default
+        delegate :coerce, :serialize, to: :type
+
+        def initialize(field:, type:)
+          @field = field
```
[the remark this answers is not in the document]
Comment (tjwp) [reply lib/avromatic/model/attributes.rb]: I don't believe there is a guiding principle. To date I think the prefix has just be used to disambiguate things that might otherwise collide.

File: lib/avromatic/model/attributes.rb [hunk 8]
```diff
@@ -79,43 +117,35 @@ def define_avro_attributes(schema, allow_optional: true)
...
-            validates(field.name, length: { is: field.type.size })
-          when :record, :array, :map, :union
-            validate_complex(field.name)
-          else
-            add_type_validation(field.name, field_class)
-          end
+            unless config.mutable
+              private("#{field.name}=")
+              define_method(:clone) { self }
+              define_method(:dup) { self }
```
[the remark this answers is not in the document]
Comment (tjwp) [reply lib/avromatic/model/attributes.rb]: I don't feel too strongly either. If it's immutable anyway having a "dup" that you can't modify doesn't seem like a big issue.

File: README.md [hunk 9]
```diff
@@ -385,8 +385,6 @@ decoder.decode(model2_message_value)
 
 The following validations are supported:
 
-- The size of the value for a fixed type field.
-- The value for an enum type field is in the declared set of values.
```
Comment (tjwp) [inline README.md]: Upgrade info to follow should make it clear that these validations are now enforced as part of coercion -- not something that has been removed.

File: lib/avromatic/model/types/fixed_type.rb [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+        def value_classes
+          VALUE_CLASSES
+        end
+
+        def coerce(input)
+          if coercible?(input)
+            input
+          else
+            raise ArgumentError.new("Could not coerce '#{input.inspect}' to a Fixed(#{size})")
```
Comment (tjwp) [inline lib/avromatic/model/types/fixed_type.rb]: What do you think about defining a `CoercionError` class within this gem instead of using `ArgumentError`?

Maybe that's already covered by the improved coercion errors that you have in mind.
